### PR TITLE
DO NOT MERGE - Allow registering multiple instrumentation hooks for the same resolved module

### PR DESCRIPTION
**Pull request by bizob2828**
## Proposed Release Notes
 * Added the ability to register instrumentation multiple hooks (onRequire, onResolved) for the same resolved moduleName.
    * This has been a limitation of the agent from the beginning.  If a customer used the api to instrument `api.instrument`, `api.instrumentDatastore`, `api.instrumentWebframework`, `api.instrumentMessages`, or `api.instrumentConglomerate`, it would override existing instrumentation hooks.  The effect was that the Node.js agent would not function as designed.  
  * **DEPRECATION NOTICE**: `shim.unwrap` and `shim.unwrapOnce` will no longer function if you attempt to unwrap an item that has been wrapped multiple times.
    * This is because since we now allow instrumenting the same module more than once, you cannot safely unwrap without breaking all registered instrumentation.  We plan to remove `shim.unwrap` and `shim.unwrapOnce` in the next major release.  

## Links
Closes NEWRELIC-7414

## Details
CI is broken because it requires that we release a new version of the agent so that https://github.com/newrelic/node-test-utilities/pull/162 can be merged and released.  Then these PRs(https://github.com/newrelic/node-newrelic-koa/pull/141, and https://github.com/newrelic/newrelic-node-apollo-server-plugin/pull/243) can be merged and then update the `test/versioned-external/external-repos.js` to point koa and apollo server to main.

Even with all those caveats this has a `DO NOT MERGE` flag because this change is high risk and we need to do more manual testing against applications that are registering multiple instrumentation hooks for a given module.  The plan is to get the k2 agent running with our versioned tests against this branch and see how well it does

### CODE REVIEW

File: lib/shim/datastore-shim.js
Comment (bizob2828) [inline lib/shim/datastore-shim.js:92]: this was done for testing. customers really couldn't do this with our APIs, I removed because it was unused and I added another arg

File: lib/shim/datastore-shim.js
Comment (bizob2828) [inline lib/shim/datastore-shim.js:545]: yea address this todo

File: lib/shim/shim.js
Comment (bizob2828) [inline lib/shim/shim.js:648]: put a comment explaining this.  since this is a proxy we don't want to propagate the wrapped symbol to the original but instead just keep it on `this` the wrapped "thing"

File: lib/shimmer.js
Comment (bizob2828) [inline lib/shimmer.js:362]: I'm not a fan of us calling this firstParty

File: lib/shimmer.js
Comment (bizob2828) [inline lib/shimmer.js:571]: hmm I should set the instrumentation[symbols.instrumented] property to true.  Even if it didn't instrument anything this is an indicator in instrumentPostLoad which tells api.instrumentedLoadedModule if it succeeded meaning every instrumentation hook was instrumented

File: test/lib/agent_helper.js
Comment (bizob2828) [inline test/lib/agent_helper.js:585]: we still use the old unwrap all in tests so I just moved this behavior to the helper

**Review by jmartin4563 — CHANGES_REQUESTED** [review]
You have a bunch of comments you left on here too, gentle reminder about those

File: lib/shim/shim.js
Comment (bizob2828) [reply lib/shim/shim.js:806]: I can't seem to figure out how to optional chain with keys that are also variables

File: lib/shim/shim.js
Comment (bizob2828) [reply lib/shim/shim.js:1025]: yes. we are making this a breaking change because of this behavior. but honestly, i doubt anyone is using this

File: test/lib/agent_helper.js
Comment (bizob2828) [reply test/lib/agent_helper.js:585]: yes there are cases where we double wrap, mostly i think in datastore shim so we still want to unwrap all in that case which is why i put this here

File: test/versioned-external/external-repos.js
Comment (bizob2828) [reply test/versioned-external/external-repos.js:38]: yea, that would be helpful. The [koa PR](https://github.com/newrelic/node-newrelic-koa/pull/141) does require to be released at same time so we should probably also semver major that.  The [apollo server PR](https://github.com/newrelic/newrelic-node-apollo-server-plugin/pull/243/files) is just a change in test behavior, so it just can be merged after we merge this one.

**Review by jmartin4563 — DISMISSED** [review]

**Review by jmartin4563 — APPROVED** [review]

**Pull request merged**
